src/OmniSharp.Roslyn.CSharp/Services/ExternalSourceServiceFactory.cs

```diff
@@ -0,0 +1,77 @@
+﻿using Microsoft.CodeAnalysis.Host;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Options;
+using OmniSharp.Roslyn.CSharp.Services.Decompilation;
+using OmniSharp.Services;
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OmniSharp.Roslyn.CSharp.Services
+{
+    [Export(typeof(ExternalSourceServiceFactory)), Shared]
+    public class ExternalSourceServiceFactory
+    {
+        private MetadataExternalSourceService _metadataExternalSourceService;
+        private DecompilationExternalSourceService _decompilationExternalSourceService;
+        private readonly IAssemblyLoader _assemblyLoader;
+        private readonly ILoggerFactory _loggerFactory;
+        private static object padlock = new object();
+
+        [ImportingConstructor]
+        public ExternalSourceServiceFactory(IAssemblyLoader assemblyLoader, ILoggerFactory loggerFactory)
+        {
+            _assemblyLoader = assemblyLoader;
+            _loggerFactory = loggerFactory;
+        }
+
+        public IExternalSourceService Create(OmniSharpOptions omniSharpOptions, HostLanguageServices hostLanguageServices)
+        {
+            var enableDecompilationSupport = omniSharpOptions.RoslynExtensionsOptions.EnableDecompilationSupport;
+
+            if (enableDecompilationSupport)
+            {
+                if (_decompilationExternalSourceService == null)
+                {
+                    lock (padlock)
+                    {
+                        if (_decompilationExternalSourceService == null)
+                        {
+                            _decompilationExternalSourceService = new DecompilationExternalSourceService(_assemblyLoader, _loggerFactory, hostLanguageServices);
+                        }
+                    }
+                }
+
+                return _decompilationExternalSourceService;
+            }
+
+            if (_metadataExternalSourceService == null)
```

Comment:
Can we solve this by having a `Lazy<MetadataExternalSourceService>` instead. Create the delegate in constructor where we have assembly loader and logger factory.